db/range_del_aggregator.cc

```diff
@@ -183,12 +183,46 @@ bool ForwardRangeDelIterator::ShouldDelete(const ParsedInternalKey& parsed) {
     PushIter(iter, parsed);
     assert(active_iters_.size() == active_seqnums_.size());
   }
+}
 
+bool ForwardRangeDelIterator::ShouldDelete(const ParsedInternalKey& parsed) {
+  AdvanceTo(parsed);
   return active_seqnums_.empty()
              ? false
              : (*active_seqnums_.begin())->seq() > parsed.sequence;
 }
 
+bool ForwardRangeDelIterator::ShouldDeleteRange(
+    const ParsedInternalKey& parsed_start, const ParsedInternalKey& parsed_end,
+    SequenceNumber seqno) {
+  ParsedInternalKey parsed_curr = parsed_start;
+  while (true) {
+    if (active_iters_.empty() || (*active_seqnums_.begin())->seq() <= seqno) {
+      return false;
+    }
+    parsed_curr = (*active_iters_.top())->end_key();
+    if (icmp_->Compare(parsed_end, parsed_curr) < 0) {
+      break;
+    }
+    AdvanceTo(parsed_curr);
+  };
+  return true;
+}
+
+RangeTombstone ForwardRangeDelIterator::GetTombstone(
+    const ParsedInternalKey& parsed, SequenceNumber seqno) {
+  AdvanceTo(parsed);
+  if (active_seqnums_.empty()) {
+    return RangeTombstone(Slice(), Slice(), 0);
```

Comment:
The previous code would set the `start_key_` to the first tombstone's end key if we are positioned before all tombstones, or the `end_key_` to the last tombstone's end key if we are positioned after all tombstones. That seems necessary for `FindKeyForward` and `FindKeyBackward` to work, as they treat an empty key as an indication the tombstone extends over the file's bounds.

Reply:
Perhaps we should introduce a `ReverseRangeDelIterator::GetTombstone()` and get the tombstone corresponding to which direction we're going. We only need to consider the case of being positioned before all keys when moving forward; likewise, we only need to consider being after all keys when moving backward. This introduces extra complexity because we'll also have to reset the cached tombstone when changing direction. Some extra complexity is expected given we are no longer operating on a global "skyline" of range tombstones.

Reply:
Done - rewrote `GetTombstone` and handled these cases.